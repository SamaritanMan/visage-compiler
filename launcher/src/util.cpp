/*
 * Copyright 2008-2009 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

#include <windows.h>

#include "util.h"

Util::Util() {
}

Util::~Util() {
}

std::string Util::evaluatePath (const std::string& visagepath, const std::string& libs) {
    std::string result = libs;
    std::string::size_type start=0;
    while ((start = result.find("${visage_home}", start)) != std::string::npos) {
        result.replace (start, 14 /* length of ${visage_home} */, visagepath);
        start += 14;
    }
    start=0;
    while ((start = result.find("/", start)) != std::string::npos) {
        result.replace (start, 1, "\\");
    }
    return result;
}

int Util::createProcess(const std::string& cmd) {
    STARTUPINFO start;
    PROCESS_INFORMATION pi;

    memset (&start, 0, sizeof (start));
    start.cb = sizeof (start);

    bool ldebug = getenv("_JAVAFX_LAUNCHER_DEBUG") != NULL;
    #ifdef DEBUG
    ldebug = true;
    #endif
    if (ldebug == true) {
        printf("Cmdline: %s\n", cmd.c_str());
    }
    if (!CreateProcess (NULL, (char*)cmd.c_str(),
                        NULL, NULL, TRUE, NORMAL_PRIORITY_CLASS,
                        NULL, 
                        NULL, // lpCurrentDirectory
                        &start,
                        &pi)) {
        fprintf (stderr, "Cannot start java.exe.");
        return EXIT_FAILURE;
    } else {
        // Wait until child process exits.
        WaitForSingleObject( pi.hProcess, INFINITE );
        unsigned long exitCode;
        GetExitCodeProcess(pi.hProcess, &exitCode);

        // Close process and thread handles. 
        CloseHandle( pi.hProcess );
        CloseHandle( pi.hThread );        
        return exitCode;
    }
    
}
