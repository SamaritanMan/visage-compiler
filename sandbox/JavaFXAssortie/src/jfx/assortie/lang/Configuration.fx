package jfx.assortie.lang;

import jfx.assortie.system.Module;
import jfx.assortie.system.Sample;

Module {
    name: "JavaFX"
    modules: [
    "jfx.assortie.lang.api.Configuration",
    ]
    samples: [
        Sample{
            name: "New"
            className: "jfx.assortie.samples.empty.Empty"
            visible: true
        }
    ]

}
