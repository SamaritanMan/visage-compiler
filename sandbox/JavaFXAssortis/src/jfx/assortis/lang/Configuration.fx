package jfx.assortis.lang;

import jfx.assortis.system.Module;
import jfx.assortis.system.Sample;

Module {
    name: "JavaFX"
    modules: [
    "jfx.assortis.lang.api.Configuration",
    "jfx.assortis.lang.animation.Configuration",
    ]
    samples: [
        Sample{
            name: "New"
            className: "jfx.assortis.samples.empty.Empty"
            visible: true
        }
    ]

}
