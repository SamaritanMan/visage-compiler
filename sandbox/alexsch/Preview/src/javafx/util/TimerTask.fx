package javafx.util;

public class TimerTask {
    var name: String;
    var time: Number;
    var action: function();
    
    public function toString () {  name } 
}  
