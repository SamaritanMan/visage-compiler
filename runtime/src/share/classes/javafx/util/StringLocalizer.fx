package javafx.util;

import java.lang.Class;
import java.lang.System;
import java.lang.Thread;
import java.lang.StackTraceElement;
import java.util.Locale;
import com.sun.javafx.runtime.util.StringLocalization;

/**
 * Utility class for localizing strings using JavaFX properties files.
 *
 * An object of this class is used to find the appropriate localized
 * strings for the given attributes.  For example,
 * <code>
 * // Object creation
 * var localizer = StringLocalizer{ key: "Hello, World!" };
 *
 * // This prints localized text for "Hello, World!" for the default locale
 * System.out.println(localizer.localizedString()); 
 *
 * // This prints localized text for "Duke" for the default locale
 * localizer.key = "Duke";
 * System.out.println(localizer.localizedString()); 
 *
 * // This prints localized text for "Duke" for the French locale
 * localizer.locale = Locale.FRENCH;
 * System.out.println(localizer.localizedString()); 
 *
 * // This prints localized text for "Duke", from 
 * // the FX properties file "foo/bar/MyResources_fr.fxproperties
 * localizer.propertiesName = "foo.bar.MyResources";
 * System.out.println(localizer.localizedString()); 
 * </code>
 */
public class StringLocalizer {

    /**
     * The string that represents the 'key' in the JavaFX properties file
     */
    public attribute key: String;

    /**
     * The locale used for determining the JavaFX properties resource
     * bundle.  If it is not explicitly specified, Locale.getDefault()
     * is used.
     */
    public attribute locale: Locale = Locale.getDefault();

    /**
     * The canonical base name of the JavaFX properties file.  It consists 
     * of the package name and the base bundle name of the JavaFX properties 
     * file.  For example, the canonical base name of 'MyResources_xx.fxproperties', 
     * where 'xx' denotes the target locale, in 'foo.bar' package is 
     * 'foo.bar.MyResources'.  If this attribute is not explicitly specified,
     * it is synthesized from the caller's package and script file name, e.g.,
     * a JavaFX Script 'Example.fx' is in 'foo.bar' package, the synthesized
     * canonical base name will be 'foo.bar.Example'.
     */
    public attribute propertiesName: String = getDefaultPropertiesName();

    /**
     * The default string returned from localizedString() method.
     */
    public attribute defaultString : String;
   
    /**
     * This function returns the localized string for the given attributes.
     * The return value is bound to each of the attributes.  If there
     * is no appropriate localized string found in JavaFX properties files,
     * this function returns 'defaultString' if it is specified, otherwise
     * 'key' is returned.
     */
    public bound function localizedString() : String {
        localizedStr;
    }

    /**
     * A static function to associate a JavaFX Script source file, or a JavaFX
     * package to a JavaFX properties file.  If an association is made, further
     * localizations in the specified source file, or source files in the 
     * specified package will be searched within the specified JavaFX properties
     * file.
     *
     * 'packageName' denotes the JavaFX package, e.g., 'foo.bar', and 'scriptFileName
     * is the file name source script file, e.g., 'Example.fx'.  If the 
     * 'scriptFileName' is null, all the JavaFX scripts in the package designated by 
     * 'packageName' are associated with the specified JavaFX properties file.
     * 'properties' denotes the canonical base name of the JavaFX properties file, e.g.,
     * 'foo.bar.MyResources'.
     */
    public static function associate(packageName: String,
                                scriptFileName: String, properties: String) : Void {
        var source: String = 
            if (scriptFileName == null) packageName 
            else packageName + "." + scriptFileName.replaceAll("\\.[fF][xX]$", "");
        StringLocalization.associate(source, properties);
    }

    private attribute localizedStr : String = bind {
        if (defaultString <> null) {
            StringLocalization.getLocalizedString(propertiesName, key, defaultString, locale);
        } else {
            StringLocalization.getLocalizedString(propertiesName, key, key, locale);
        }
    }

    private function getDefaultPropertiesName() : String {
        var elements  = Thread.currentThread().getStackTrace();
        var elem: StackTraceElement;
        var className: String;
        var foundMe: Boolean = false;

//      JFXC-555
//      for (elem in elements) {
        for (i in [0..<sizeof elements]) {
            elem = elements[i];
            className = elem.getClassName();

            if (className.startsWith("javafx.util.StringLocalizer")) {
                foundMe = true;
            } else if (foundMe) {
                // this should be the caller's stack
                break;
            }
        }

        var pkgName = className.replaceAll("\\.?[^\\.]+$", "");
        StringLocalization.getPropertiesName(
            (if (pkgName <> "") pkgName + "." else "") +
            elem.getFileName().replaceAll("\\.[fF][xX]$", ""));
    }
}
