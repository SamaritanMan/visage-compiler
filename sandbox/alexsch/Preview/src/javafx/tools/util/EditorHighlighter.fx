package javafx.tools.util;


import javafx.tools.*;


import java.lang.System;
import javax.swing.text.JTextComponent;

public class EditorHighlighter{
    
    
    public attribute component: JTextComponent;

    public static attribute KEY_WORDS:String[] =  ["import"];

    
    public function clear(){
        var highlight = component.getHighlighter();
        highlight.removeAllHighlights();
    }
    
    public function highlightKeyWords(){
        //System.out.println("[highlight] keywords");
        var hilite = component.getHighlighter();
        //var painter = new javax.swing.text.DefaultHighlighter.DefaultHighlightPainter(new java.awt.Color(188,207,246));
        var painter = new javax.swing.text.DefaultHighlighter.DefaultHighlightPainter(new java.awt.Color(222,236,242));

        var content = component.getText();
        //System.out.println("  content:\n {content}");
        
        for(keyWord in KEY_WORDS){
            //System.out.println("  keyword: '{keyWord}'");
            
            
            var index = content.indexOf(keyWord, 0);
            
            while (0 <= index){
                var end = index + keyWord.length();
                //System.out.println("  positions: {index} - {end}");
                hilite.addHighlight(index, end, painter);
                index = content.indexOf(keyWord, end);
                
            }
            
            
        }
       
    }
    
    public function highlightErrors(diagnosticMessages: DiagnosticMessage[]){
        for(diagnosticMessage in diagnosticMessages){
            highlightError(diagnosticMessage);
        }
    }
    
    public function highlightError(diagnosticMessage: DiagnosticMessage){
        var highlight = component.getHighlighter();
        var painter = new javax.swing.text.DefaultHighlighter.DefaultHighlightPainter(new java.awt.Color(254,134,126));
        var startPosition = diagnosticMessage.startPosition;
        var endPosition = diagnosticMessage.endPosition;
        //System.out.println("\n[highliter] error: [{startPosition} - {endPosition}]");
        if(startPosition == endPosition){
            startPosition -= 3;
            endPosition;
        }
        
        highlight.addHighlight(startPosition, endPosition, painter);
        
    }
    
}

