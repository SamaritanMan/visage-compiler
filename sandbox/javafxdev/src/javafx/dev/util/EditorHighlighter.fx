package javafx.dev.util;

/**
 * @author Alexandr Scherbatiy, alexsch@dev.java.net
 */

import javafx.dev.*;

import java.lang.System;
import javax.swing.text.JTextComponent;

public class EditorHighlighter{
    
    
    public var component: JTextComponent;

    public static var KEY_WORDS:String[] =  ["import"];

    
    public function clear(){
        var highlight = component.getHighlighter();
        highlight.removeAllHighlights();
    }
    
    public function highlightKeyWords(){
        //System.out.println("[highlight] keywords");
        var hilite = component.getHighlighter();
        //var painter = new javax.swing.text.DefaultHighlighter.DefaultHighlightPainter(new java.awt.Color(188,207,246));
        var painter = new javax.swing.text.DefaultHighlighter.DefaultHighlightPainter(new java.awt.Color(222,236,242));

        var content = component.getText();
        //System.out.println("  content:\n {content}");
        
        for(keyWord in KEY_WORDS){
            //System.out.println("  keyword: '{keyWord}'");
            
            
            var index = content.indexOf(keyWord, 0);
            
            while (0 <= index){
                var end = index + keyWord.length();
                //System.out.println("  positions: {index} - {end}");
                hilite.addHighlight(index, end, painter);
                index = content.indexOf(keyWord, end);
                
            }
            
            
        }
       
    }
    
    public function highlightErrors(ErrorMessages: ErrorMessage[]){
        for(errorMessage in ErrorMessages){
            highlightError(errorMessage);
        }
    }
    
    public function highlightError(errorMessage: ErrorMessage){
        var highlight = component.getHighlighter();
        var painter = new javax.swing.text.DefaultHighlighter.DefaultHighlightPainter(new java.awt.Color(254,134,126));
        var startPosition = errorMessage.startPosition;
        var endPosition = errorMessage.endPosition;
        //System.out.println("\n[highliter] error: [{startPosition} - {endPosition}]");
        if(startPosition == endPosition){
            startPosition -= 3;
            endPosition;
        }
        
        highlight.addHighlight(startPosition, endPosition, painter);
        
    }
    
}

