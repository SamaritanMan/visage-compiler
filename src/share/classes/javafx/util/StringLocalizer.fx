package javafx.util;

import java.lang.Class;
import java.lang.System;
import java.lang.Thread;
import java.lang.StackTraceElement;
import java.util.Locale;
import com.sun.javafx.runtime.util.StringLocalization;

/**
 * Utility class for localizing strings using JavaFX properties files.
 *
 * An object of this class is used to find the appropriate localized
 * strings for the given attributes.  For example,
 * <code><pre>
 * // Object creation
 * var localizer = StringLocalizer{ key: "Hello, World!" };
 *
 * // This prints localized text for "Hello, World!" for the default locale
 * System.out.println(localizer.localizedString); 
 *
 * // This prints localized text for "Duke" for the default locale
 * localizer.key = "Duke";
 * System.out.println(localizer.localizedString); 
 * </pre></code>
 *
 * By default, translations are retrieved from the JavaFX properties file
 * which has the same package and file name as the caller's script file.  For
 * example, a call from the script <code>foo.bar.Example</code> will search
 * translated strings from <code>foo/bar/Example_xx.fxproperties</code> where 'xx'
 * represents the locale. By calling <code>associate()/dissociate()</code> functions,
 * different JavaFX properties files can be used for the translation.  For example,
 * <pre><code>
 * StringLocalizer.associate("foo.bar.resources.MyResources", "foo.bar");
 * </code></pre>
 * After this call, all the translations from JavaFX script files in 
 * <code>foo.bar</code> package are searched in 
 * <code>foo/bar/resources/MyResources_xx.fxproperties</code>, where 'xx' denotes
 * the default locale.
 *
 * @needsreview
 */

public class StringLocalizer {

    /**
     * The string that represents the 'key' in the JavaFX properties file
     *
     * @needsreview
     */
    public attribute key: String;

    /**
     * The locale used for determining the JavaFX properties resource
     * bundle.  If it is not explicitly specified, Locale.getDefault()
     * is used.  
     * Note: Made this private for now, as there is no Locale class in some
     * platforms (e.g., CLDC/MIDP)
     */
    private attribute locale: Locale = Locale.getDefault();

    /**
     * The canonical base name of the JavaFX properties file.  It consists 
     * of the package name and the base bundle name of the JavaFX properties 
     * file.  For example, the canonical base name of 'MyResources_xx.fxproperties', 
     * where 'xx' denotes the locale, in 'foo.bar' package is 
     * 'foo.bar.MyResources'.  If this attribute is not explicitly specified,
     * it is synthesized from the caller's package and script file name, e.g.,
     * a JavaFX Script 'Example.fx' is in 'foo.bar' package, the synthesized
     * canonical base name will be 'foo.bar.Example'.
     *
     * @needsreview
     * @defaultvalue the caller's script file name
     */
    private attribute propertiesName: String = getDefaultPropertiesName();

    /**
     * The default string for the <code>localizedString</code>.
     *
     * @needsreview
     */
    public attribute defaultString : String;
   
    /**
     * The localized string for the given attributes.  If there
     * is no appropriate localized string found in JavaFX properties files,
     * this attribute will have the value in <code>defaultString</code> if 
     * it is specified, otherwise it will have the value in <code>key</code>.
     *
     * @needsreview
     * @readonly
     */
    public attribute localizedString : String = bind {
        if ("".equals(defaultString)) {
            StringLocalization.getLocalizedString(propertiesName, key, key, locale);
        } else {
            StringLocalization.getLocalizedString(propertiesName, key, defaultString, locale);
        }
    }

    private function getDefaultPropertiesName() : String {
        var elements  = Thread.currentThread().getStackTrace();
        var elem: StackTraceElement;
        var className: String;
        var foundMe: Boolean = false;

        for (i in [0..<sizeof elements]) {
            elem = elements[i];
            className = elem.getClassName();

            if (className.startsWith("javafx.util.StringLocalizer")) {
                foundMe = true;
            } else if (foundMe) {
                // this should be the caller's stack
                break;
            }
        }

        var pkgName = className.replaceAll("\\.?[^\\.]+$", "");
        var fileName = elem.getFileName().replaceAll("\\.[fF][xX]$", "");
        StringLocalization.getPropertiesName("{pkgName}/{fileName}");
    }
}


/**
 * Associates a JavaFX
 * package with a JavaFX properties file.  If an association is made, further
 * localizations in the source script files in the 
 * specified package will be searched within the specified JavaFX properties
 * file.
 *
 * @needsreview
 * @param properties denotes the canonical base name of the JavaFX properties 
 *     file, e.g., 'foo.bar.MyResources'.
 * @param packageName denotes the JavaFX package name, e.g., 'foo.bar'.
 */
public function associate(properties: String, packageName: String) : Void {
    associate(properties, packageName, "");
}

/**
 * Associates a JavaFX Script source file
 * with a JavaFX properties file.  If an association is made, further
 * localizations in the specified source script file
 * will be searched within the specified JavaFX properties
 * file.
 *
 * @needsreview
 * @param properties denotes the canonical base name of the JavaFX properties 
 *     file, e.g., 'foo.bar.MyResources'.
 * @param packageName denotes the JavaFX package name of the source script file, 
 *     e.g., 'foo.bar'.
 * @param scriptFileName the file name of the source script file, e.g., 'Example.fx'.
 */
public function associate(properties: String, packageName: String,
                            scriptFileName: String) : Void {
    var source: String = if (scriptFileName == "") {
            packageName;
        } else {
            "{packageName}/{scriptFileName.replaceAll("\\.[fF][xX]$", "")}";
        };
    StringLocalization.associate(source, properties);
}

/**
 * Dissociates a JavaFX
 * package from any JavaFX properties file.
 *
 * @needsreview
 * @param packageName denotes the JavaFX package name, e.g., 'foo.bar'.
 */
public function dissociate(packageName: String) : Void {
    dissociate(packageName, "");
}

/**
 * Dissociates a JavaFX Script source file
 * from any JavaFX properties file.
 *
 * @needsreview
 * @param packageName denotes the JavaFX package name, e.g., 'foo.bar'.
 * @param scriptFileName the file name of the source script file, e.g., 'Example.fx'.
 */
public function dissociate(packageName: String,
                            scriptFileName: String) : Void {
    var source: String = if (scriptFileName == "") {
            packageName;
        } else {
            "{packageName}/{scriptFileName.replaceAll("\\.[fF][xX]$", "")}";
        };
    StringLocalization.dissociate(source);
}
