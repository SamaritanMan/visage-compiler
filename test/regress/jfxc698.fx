/* 
 * test for the generic localization functionality with FX properties files
 *
 * @test
 * @run
 */

import java.lang.System;
import java.util.*;

import javafx.util.StringLocalizer;

// save the default locale for testing
var curLoc = Locale.getDefault();

// set the default locale to Japan
Locale.setDefault(Locale.JAPAN);

// create localizer object literal with defaults
var localizer = StringLocalizer { key: "source original string" };

// This prints translated string in jfxc698_ja.fxproperties
System.out.println(localizer.localizedString());

// This prints translated string in jfxc698_fr.fxproperties
localizer.locale = Locale.FRENCH;
System.out.println(localizer.localizedString());

// This prints translated string in foo/bar/FXPropTestResources_ja.fxproperties
localizer.locale = Locale.JAPANESE;
localizer.propertiesName = 'foo/bar/FXPropTestResources';
System.out.println(localizer.localizedString());

// create another localizer object literal with defaults
var localizer2 = StringLocalizer { key: "EXPLICIT_KEY", defaultString: "default string" };

// This prints translated string in jfxc698_ja.fxproperties
System.out.println(localizer2.localizedString());

// This prints the 'default string'
localizer2.key = "EXPLICIT_KEY2";
System.out.println(localizer2.localizedString());

// restore the default locale
Locale.setDefault(curLoc);
