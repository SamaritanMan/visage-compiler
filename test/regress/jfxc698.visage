/* 
 * test for the generic localization functionality with FX properties files
 *
 * @test
 * @run
 */

import java.lang.System;
import java.util.*;

import visage.util.StringLocalizer;

// save the default locale for testing
var curLoc = Locale.getDefault();

try {
    // set the default locale to Japan
    Locale.setDefault(Locale.JAPAN);

    // create localizer object literal with defaults
    var localizer = StringLocalizer { key: "source original string" };

    // This prints translated string in jfxc698_ja.fxproperties
    System.out.println(localizer.localizedString);

    // This prints translated string in foo/bar/FXPropTestResources_ja.fxproperties
    StringLocalizer.associate("foo.bar.FXPropTestResources", "", "jfxc698.visage");
    localizer = StringLocalizer { key: "source original string" };
    System.out.println(localizer.localizedString);

    // create another localizer object literal with defaults
    StringLocalizer.dissociate("", "jfxc698.visage");
    var localizer2 = StringLocalizer { key: "EXPLICIT_KEY", defaultString: "default string" };

    // This prints translated string in jfxc698_ja.fxproperties
    System.out.println(localizer2.localizedString);

    // This prints the 'default string'
    localizer2.key = "EXPLICIT_KEY2";
    System.out.println(localizer2.localizedString);

} finally {
    // restore the default locale
    Locale.setDefault(curLoc);
}
