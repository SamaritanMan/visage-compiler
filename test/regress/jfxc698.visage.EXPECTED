translated string in jfxc698_ja.fxproperties
translated string in foo/bar/FXPropTestResources_ja.fxproperties
translated EXPLICIT_KEY in jfxc698_ja.fxproperties
default string
